# Symmetry Adapted PDF Analysis - Running SAPA









First, import the sapa library

In [1]:
import sapa as s

Next, we need to create a SAPA object. This imports all the data from a CIF.

In [4]:
struc = s.sapa('BaTiO3_iso.cif')

We can see the list of "methods" the SAPA object has:

In [5]:
dir(struc)

['__class__',
 '__delattr__',
 '__dict__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__le__',
 '__lt__',
 '__module__',
 '__ne__',
 '__new__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__setattr__',
 '__sizeof__',
 '__str__',
 '__subclasshook__',
 '__weakref__',
 'atom_site_Wyckoff_label',
 'atom_site_fract_symmform',
 'atom_site_fract_x',
 'atom_site_fract_y',
 'atom_site_fract_z',
 'atom_site_label',
 'atom_site_occupancy',
 'atom_site_symmetry_multiplicity',
 'atom_site_type_symbol',
 'cell_angle_alpha',
 'cell_angle_beta',
 'cell_angle_gamma',
 'cell_length_a',
 'cell_length_b',
 'cell_length_c',
 'cell_volume',
 'cleanup',
 'constraint_array',
 'constraints_out',
 'create_hdf5',
 'create_hdf5_single',
 'execute',
 'execute_occ',
 'execute_single',
 'file_check',
 'get_indices',
 'idformat',
 'instrument_pars',
 'irrep_list',
 'iso_coordinate_formula',
 'iso_coordinate_label',

A lot of these are variables imported from the CIF itself. For example:

In [6]:
print(struc.cell_length_a)
print(struc.cell_length_b)
print(struc.cell_length_c)

8.07084
8.07084
8.07084


This gives the lattice parameters of the supercell we chose when we generated the CIF with ISODISTORT.

Other methods operate on the information provided in the CIF. For example:

In [7]:
struc.irrep_list()

['GM4-',
 'R4-',
 'X3-',
 'X5-',
 'M3-',
 'M5-',
 'R5+',
 'X1+',
 'X5+',
 'M2-',
 'GM5-',
 'R2-',
 'R3-',
 'R5-',
 'X2+',
 'M1+',
 'M2+',
 'M3+',
 'M4+',
 'M5+']

This provides a list of irreps contained within the CIF. 

We are going to use the "write_inp" method to make an input file to perform symmetry adapted PDF analysis. The required arguments takes are (in order):

sample: 
A "sample" string to remind you what you are running. This is often just the material you are analysing. This is repeated in the output files SAPA produces.

filenameformat: 
This is a string to allow Topas to iterate over the data files. This includes any path information - SAPA assumes the data is in the working directory, so you need to specify if it isn't. Your data file names should be in a format to allow for iteration, i.e. they have a consistent naming convention such as "BaTiO3_15K.xy", "BaTiO3_150K.xy". The temperatures here can be iterated over, and to tell Topas that, replace the temperature with ##n## (filenameformat = "BaTiO3_##n##K.xy")

isneutron:
Set this to True if you are using neutron data, or False for X-ray data. For this example with BaTiO3, we are using neutron data.

qmax:
The maximum value of Q used to generate the PDF data. For this data, qmax = 40.

dq:
The Q-space dampening parameter. For this data, dq = 0.033

startx, finishx:
The r-range to start and end the refinement at. For SAPA, most of the information is contained within the low-r region. For this data, set startx = 1.7 and finishx = 10.0

lattice:
This is a flag to allow lattice parameters to refine as they should for a given lattice type. This can be "C", "T", "O", "R", "H", "TC" or "M", for cubic, tetragonal, orthorhombic, rhombohedral, hexagonal, triclinic and monoclinic, respectively. For now, use lattice = "C".

There are further optional arguments that can be specified. The relevant ones for today are:

cycles:
The number of cycles to perform for each irrep. The default is 300, but for time purposes, lets lower this to 100.

filename:
The default is "batch_modes.inp", but this can be changed. Obviously, it still requires the .inp file extension, whatever the name is. 

In [8]:
struc.write_inp(sample="BaTiO3", filenameformat="BaTiO3_##n##K.xy", isneutron=True, qmax=40.0, dq=0.033, startx=1.7, finishx=10.0, lattice="C", cycles=100)

Creating input file...
120
120
...batch_modes.inp written


This has printed the number of modes in the input file (120 in this case), and has indicated an input file has been written. This can be viewed using jEdit, and we can go through the components of that input file later, time providing.

Now, we set the temperatures of our data files. (NB: these don't have to be temperatures, it is simply something to iterate over). This can be specified as a list (as below) if you don't have a regular interval between data points. Often at I15-1, we collect data at regular intervals, so we could use temps = range(start, finish + interval, interval)

In [9]:
temps = [15, 150, 210, 250, 293, 350, 410, 500]

And now, we can set the code to be run. For each temperature and irrep, this tells Topas to execute the input file with the correct modes activated. 

In [10]:
struc.execute(temps)

Finding Topas executables...
Creating Monitoring File...
Executing...
Executing irrep GM4- for temp 15
Completed irrep GM4- for temp 15. Process took 123.49915361404419s (2.0583192269007364 minutes)
Estimated 5.728988514873716 hours to completion.
Executing irrep R4- for temp 15
Completed irrep R4- for temp 15. Process took 127.39775681495667s (2.1232959469159445 minutes)
Estimated 5.784567657113075 hours to completion.
Executing irrep X3- for temp 15
Completed irrep X3- for temp 15. Process took 143.36500883102417s (2.389416813850403 minutes)
Estimated 6.023445988694827 hours to completion.
Executing irrep X5- for temp 15
Completed irrep X5- for temp 15. Process took 155.01531505584717s (2.5835885842641195 minutes)
Estimated 6.255657390819655 hours to completion.
Executing irrep M3- for temp 15
Completed irrep M3- for temp 15. Process took 81.90768074989319s (1.36512801249822 minutes)
Estimated 5.715730064206654 hours to completion.
Executing irrep M5- for temp 15
Completed irrep M5- 

Completed irrep M3- for temp 210. Process took 80.19543361663818s (1.3365905602773032 minutes)
Estimated 3.977064160869477 hours to completion.
Executing irrep M5- for temp 210
Completed irrep M5- for temp 210. Process took 239.31595277786255s (3.9885992129643757 minutes)
Estimated 4.028216748767429 hours to completion.
Executing irrep R5+ for temp 210
Completed irrep R5+ for temp 210. Process took 97.893479347229s (1.6315579891204834 minutes)
Estimated 3.979164093600379 hours to completion.
Executing irrep X1+ for temp 210
Completed irrep X1+ for temp 210. Process took 143.6057412624359s (2.3934290210405984 minutes)
Estimated 3.9609527665694557 hours to completion.
Executing irrep X5+ for temp 210
Completed irrep X5+ for temp 210. Process took 252.71919226646423s (4.211986537774404 minutes)
Estimated 4.011424358054703 hours to completion.
Executing irrep M2- for temp 210
Completed irrep M2- for temp 210. Process took 154.88658952713013s (2.5814431587855022 minutes)
Estimated 3.9966319

Completed irrep M2- for temp 293. Process took 167.13168692588806s (2.7855281154314677 minutes)
Estimated 2.458249845945807 hours to completion.
Executing irrep GM5- for temp 293
Completed irrep GM5- for temp 293. Process took 143.776029586792s (2.3962671597798666 minutes)
Estimated 2.430192654741438 hours to completion.
Executing irrep R2- for temp 293
Completed irrep R2- for temp 293. Process took 53.23825240135193s (0.8873042066891988 minutes)
Estimated 2.3830259142319363 hours to completion.
Executing irrep R3- for temp 293
Completed irrep R3- for temp 293. Process took 65.23947286605835s (1.0873245477676392 minutes)
Estimated 2.33896687899122 hours to completion.
Executing irrep R5- for temp 293
Completed irrep R5- for temp 293. Process took 72.86855101585388s (1.2144758502642314 minutes)
Estimated 2.2969508754355568 hours to completion.
Executing irrep X2+ for temp 293
Completed irrep X2+ for temp 293. Process took 82.56987977027893s (1.3761646628379822 minutes)
Estimated 2.25725

Completed irrep R5- for temp 410. Process took 80.98551106452942s (1.349758517742157 minutes)
Estimated 0.9099289445612166 hours to completion.
Executing irrep X2+ for temp 410
Completed irrep X2+ for temp 410. Process took 84.12205076217651s (1.4020341793696085 minutes)
Estimated 0.8756831411097912 hours to completion.
Executing irrep M1+ for temp 410
Completed irrep M1+ for temp 410. Process took 70.18018126487732s (1.1696696877479553 minutes)
Estimated 0.840881463394665 hours to completion.
Executing irrep M2+ for temp 410
Completed irrep M2+ for temp 410. Process took 82.9945056438446s (1.3832417607307435 minutes)
Estimated 0.8069161684038848 hours to completion.
Executing irrep M3+ for temp 410
Completed irrep M3+ for temp 410. Process took 59.146318197250366s (0.9857719699541728 minutes)
Estimated 0.7719983360171319 hours to completion.
Executing irrep M4+ for temp 410
Completed irrep M4+ for temp 410. Process took 74.34636664390564s (1.2391061107317607 minutes)
Estimated 0.73800

Occasionally, things can interrupt the refinement process, or cause it to hang (e.g. locking your computer). The execute script writes a file to keep track of the temperature and irrep combinations it has and hasn't ran so far, and this can be used to restart the refinement.

In [11]:
#struc.restart() 

## Part 2 - Post Execution

Sometimes something can go wrong with an individual irrep/temp combo, but not interrupt the overall SAPA run. An indicator of this is that the ycalc and yobs files for these runs haven't been generated. Missing files causes problems with the HDF5 file creation. The following script checks that all the expected files are there, and if any are found to be missing, it re-runs them.

In [12]:
struc.file_check(temps, cycles=100, verbose=False) #re-run a couple times to make sure

Files checked, found no issues!


SAPA generates a large amount of files, which can be quite cumbersome to deal with. To make dealing with SAPA data more straightforward, we place all the data in these files into a structured file format (HDF5). The file is structured in the same way for every SAPA run, meaning analysis is more repeatable.

In [13]:
struc.create_hdf5(temps)

We are still left with the large number of files however. The following script removes them and (optionally) compresses them into an archive.

In [14]:
struc.cleanup() # optionally set create_zip = False

Creating zip file...
...Zip file written.
Deleting .txt files...
...Clean up complete.
